### PR TITLE
Calling the emergency shuttle now requires a physical button

### PR DESCRIPTION
**Pull request by Arrow768**
* The emergency shuttle can now be called by using a physical button in a protected case.
* The option to call the emergency shuttle has been removed from the command and communications app.
* Also adds a framework for physical buttons in a protected case.
* The emergency shuttle button needs to be mapped in.

### CODE REVIEW

File: code/game/machinery/case_button.dm [hunk 1]
```diff
@@ -20,21 +21,26 @@
...
     update_icon()
 
 /obj/machinery/case_button/Destroy()
     QDEL_NULL(listener)
+    return ..()
 
 /obj/machinery/case_button/attackby(obj/item/weapon/W, mob/user)
     if(istype(W, /obj/item/weapon/card))
-        if(access_keycard_auth in W.GetAccess())
+        if(cover_access && cover_access in W.GetAccess())
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Please use `req_access` and `allowed()`.

File: code/game/machinery/case_button.dm [hunk 2]
```diff
@@ -20,21 +21,26 @@
...
+        else if(!cover_access)
             covered = !covered //Enable / Disable the forcefield
+        update_use_power(covered + 1) //Update the power usage
     else
         if(covered && (stat & NOPOWER)) //Only bounce off if its powered (i.e. shield active)
-            user.visible_message("<span class='danger'>[src] has been hit by [user] with [W], but it bounces off the forcefield</span>","<span class='danger'>You hit [src] with [W], but it bounces off the forcefield</span>","You hear something boucing off a forcefield")
-        else
             ..()
+        else
+            user.visible_message("<span class='danger'>[src] has been hit by [user] with [W], but it bounces off the forcefield</span>","<span class='danger'>You hit [src] with [W], but it bounces off the forcefield</span>","You hear something boucing off a forcefield")
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Periods.

File: code/game/machinery/case_button.dm [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            return ..()
+        last_toggle_time = world.time
+        if(!active)
+            if(activate(user))
+                for(var/obj/machinery/case_button/cb in get_listeners_by_type(button_type,/obj/machinery/case_button))
+                    cb.active = 1
+                    cb.update_icon()
+        else
+            if(deactivate(user))
+                for(var/obj/machinery/case_button/cb in get_listeners_by_type(button_type,/obj/machinery/case_button))
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Typeless for here - GLBT already enforces the type.

File: code/game/machinery/case_button.dm [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    cb.active = 0
+                    cb.update_icon()
+    else
+        ..()
+    return
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Trailing return.

File: code/game/machinery/case_button.dm [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+/obj/machinery/case_button/power_change()
+    . = ..()
+    update_icon()
+    return
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Trailing return.

File: code/game/machinery/case_button.dm [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+/obj/machinery/case_button/update_icon()
+    cut_overlays()
+    if(stat & NOPOWER)
+        update_use_power(0)
+        add_overlay("b[button]d") //Add the deactivated button overlay
+        add_overlay("g[cover]d") //Add the deactivated cover overlay
+        return
+    add_overlay("b[button][active]") //Add the button as overlay
+    add_overlay("g[cover][covered]") //Add the glass/shield overlay
+    return
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Trailing return.

File: code/game/machinery/case_button.dm [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    else
+        if(covered && (stat & NOPOWER)) //Only bounce off if its powered (i.e. shield active)
+            ..()
+        else
+            user.visible_message("<span class='danger'>[src] has been hit by [user] with [W], but it bounces off the forcefield.</span>","<span class='danger'>You hit [src] with [W], but it bounces off the forcefield.</span>","You hear something boucing off a forcefield.")
+    update_icon()
+    return
+
+/obj/machinery/case_button/attack_hand(mob/user as mob)
+    if(covered == 0)
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: `if (!covered)`

File: code/game/machinery/case_button.dm [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        //Spam Check
+        if((last_toggle_time + timeout) > world.time)
+            user.visible_message("<span class='notice'>\The [user] presses the button, but nothing happens.</span>","<span class='notice'>You press the button, but it is not responding.</span>","You hear something being pressed.")
+            return ..()
+        last_toggle_time = world.time
+        if(!active)
+            if(activate(user))
+                for(var/obj/machinery/case_button/cb in get_listeners_by_type(button_type,/obj/machinery/case_button))
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: Typeless for here too.

File: code/game/machinery/case_button.dm [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    listener = new(button_type, src)
+    update_icon()
+
+/obj/machinery/case_button/Destroy()
+    QDEL_NULL(listener)
+    return ..()
+
+/obj/machinery/case_button/attackby(obj/item/weapon/W, mob/user)
+    if(istype(W, /obj/item/weapon/card))
+        if(src.allowed(usr))
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: `user`, not `usr`.

File: code/game/machinery/case_button.dm [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+//Deactivate Button - Needs ro return 1 for the activation to be successful
+/obj/machinery/case_button/proc/deactivate(mob/user)
+    user.visible_message("<span class='notice'>\The [user] resets the button.</span>","<span class='notice'>You reset the button.</span>","You hear something being pressed.")
+    return 1
+
+
+
+
+/obj/machinery/case_button/shuttle
+    name = "Emergency Shuttle Button"
```
Comment (Lohikar) [inline code/game/machinery/case_button.dm]: This should probably be either lowercase or `\improper`.

File: code/game/machinery/case_button.dm (hunk 4 above)
Comment (Arrow768) [reply code/game/machinery/case_button.dm]: Style question. Wont be changed

File: code/game/machinery/case_button.dm (hunk 5 above)
Comment (Arrow768) [reply code/game/machinery/case_button.dm]: Style question. Wont be changed

File: code/game/machinery/case_button.dm (hunk 6 above)
Comment (Arrow768) [reply code/game/machinery/case_button.dm]: Style question. Wont be changed

**Comment by skull132** [comment]
Juani is doing the mapping, so I'll merge this now.

**Pull request merged**